tls-client/main.cpp

```diff
@@ -256,22 +261,31 @@ class HelloHTTPS {
 
 
         /* Read data out of the socket */
+        int read = 0;
         do {
-            ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _buffer, sizeof(_buffer));
-        } while (ret != 0 && (ret == MBEDTLS_ERR_SSL_WANT_READ ||
-                ret == MBEDTLS_ERR_SSL_WANT_WRITE));
+            ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _buffer + read, 
+                                   sizeof(_buffer) - 1);
+            if (ret > 0)
+              read += ret;
+            else if (ret != MBEDTLS_ERR_SSL_WANT_READ &&
```

Comment:
Because this condition is not sufficient as the loop condition. ```ret``` can have positive values. But I can squeeze the new condition in while if you want.